User parameter value json fix

adam-james-v opened the pull request:
The `json-in` and `json-out` functions used for toucan model transforms do not perform any string escaping logic. This
makes sense as we don't want to make assumptions about the shape of data flowing in/out of the db. But, this did mean
that for the User paramter values table, string values were causing an error to be logged. This doesn't break
anything, as the function will still correctly return the string, but it can clutter up logs.

So, this PR creates a json-out version that doesn't log this error, to keep unneeded errors from cluttering the logs

adam-james-v committed:
UserParameterValue transforms wrapped to properly escape string vals

The `json-in` and `json-out` functions used for toucan model transforms do not perform any string escaping logic. This
makes sense as we don't want to make assumptions about the shape of data flowing in/out of the db. But, this did mean
that for the User paramter values table, string values were causing an error to be logged. This doesn't break
anything, as the function will still correctly return the string, but it can clutter up logs.

So, this PR wraps all incoming values in a map and unwraps it for outgoing values.

adam-james-v committed:
Add migration to wrap existing user param values with ::wrapper key

Since the in/out transform for UserParameterValues is updated, we need to migrate any existing values to have this wrapping.

dpsutton commented:
Small thing: you use `keep` when you are unconditionally returning a map. Just use `map` if you always get a result.

The big thing is that I don't think you have looked at the values on stats. 

For instance, one of the values is `["open"]`. What does your migration do to this value? Running in the repl shows very poor results

```clojure
interface=> (let [f (fn [{:keys [id value]}]
                      (let [parsed-value (try
                                           (json/parse-string value keyword)
                                           (catch Throwable e
                                             (format "\"%s\"" value)))]
                        {:id    id
                         :value (format "{\"metabase.models.user-parameter-value/wrapper\":%s}" parsed-value)}))]
              (f {:id 1 :value "[\"open\"]"}))
{:id 1,
 :value "{\"metabase.models.user-parameter-value/wrapper\":clojure.lang.LazySeq@342669}"}
```

adam-james-v committed:
Simplify the solution to the problem.

Since we're already getting what we need from the json-in/json-out *except* that it's logging a parse error, we can
create a json-out that works the same way (tries to parse and returns the string as-is if it fails) without logging an
error at all.

adam-james-v committed:
take away the arg and keywordize inside the json-out fn

adam-james-v committed:
remove irrelevant comment

Signed-off-by: Adam James <adam.vermeer2@gmail.com>

adam-james-v committed:
Add a little more test coverage for different value types

dpsutton reviewed (approved):
A few things
1. I think the change to the migration file should be removed. I think it's a vestige of the previous approach and we don't need to modify migrations in this change at all
2. I think you should put a comment about the data shape in here
3. I think we can do tests without the `mt/with-empty-db`. As is, these tests run in about a second. I've got a modified version here at 40 msecs or so.


```clojure
(deftest user-parameter-value-crud-test
  (let [user-id        (mt/user->id :rasta)
        original-count (t2/count :model/UserParameterValue)
        param-name     (str (random-uuid))
        value!         (fn
                         ([] (->> param-name
                                  (t2/select-one :model/UserParameterValue :user_id user-id :parameter_id)
                                  :value))
                         ([v] (upv/upsert! user-id param-name v)))]
    (try
      (doseq [[test-str value-in value-out value-update value-update-out] [["string" "A" "A" "B" "B"]
                                                                           ["key" :A "A" :B "B"]
                                                                           ["vectors"
                                                                            ["A" "B" "C"] ["A" "B" "C"]
                                                                            ["A" "B" "C" "D"] ["A" "B" "C" "D"]]]]
        (testing (format "User Parameter Value for %s values" test-str)
          (testing (format "Upsert creates new user parameter value entry if the param_id user_id pair doesn't exist")
            (value! value-in)
            (is (= (inc original-count) (t2/count :model/UserParameterValue)))
            (is (= value-out (value!))))

          (testing "Upsert updates user parameter value entry if the param_id user_id pair already exists"
            (value! value-update)
            (is (= (inc original-count) (t2/count :model/UserParameterValue)))
            (is (= value-update-out (value!))))

          (testing "Upsert deletes user parameter value entry if value is `nil`."
            (value! nil)
            (is (= original-count (count (t2/select :model/UserParameterValue))))
            (is (= nil (value!))))))
      (finally
        (t2/delete! :model/UserParameterValue :parameter_id param-name)))))
```

dpsutton commented:
I suspect it used `mt/with-empty-db` in the test since it used a static param name. Left over cruft would make it flake. If we just use a unique param name we should be good and clean up after ourselves in either case.

adam-james-v committed:
faster test thanks to Dan!

adam-james-v committed:
Add a comment about the data being tested
